### PR TITLE
Add 0D `F` DataLayout

### PR DESCRIPTION
**Pull request by jb-mackay**
Introduces `DataF` `DataLayout` for field values at a single point.

Todo: extend broadcasting machinery to integrate with other DataLayouts.

This is the first step towards the `PointSpace`, described [here](https://github.com/CliMA/ClimaCore.jl/pull/325#issuecomment-1083349501)

### CODE REVIEW

File: src/DataLayouts/DataLayouts.jl [hunk 1]
```diff
@@ -411,6 +415,100 @@ end
     IFH{SS, Ni}(dataview)
 end
 
+# ======================
+# Data0D DataLayout
+# ======================
+
+Base.size(data::Data0D) = (1, 1, 1, 1, 1)
+
+struct F{S, A} <: Data0D{S}
```
Comment (simonbyrne) [inline src/DataLayouts/DataLayouts.jl:424]: Maybe call it `DataF`? I realize it breaks the convention, but calling it `F` seems like a great way to cause confusion.

File: src/DataLayouts/DataLayouts.jl [hunk 2]
```diff
@@ -411,6 +415,100 @@ end
...
+#                  throw(BoundsError(data, (v,)))
+#     dataview = @inbounds view(parent(data), v, :)
+#     @inbounds get_struct(dataview, S)
+# end
+@inline function Base.getindex(data::F{S}, v::Integer) where {S}
+    @boundscheck 1 <= v <= size(parent(data), 1) ||
+                 throw(BoundsError(data, (v,)))
+    dataview = @inbounds view(parent(data), v, :)
+    @inbounds get_struct(dataview, S)
+end
```
Comment (simonbyrne) [inline src/DataLayouts/DataLayouts.jl:475]: The folllowing should be sufficient:
```suggestion
@inline function Base.getindex(data::F{S) where {S}
    @inbounds get_struct(parent(data), S)
end
```

File: test/DataLayouts/data0d.jl [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+@testset "F" begin
+    for FT in TestFloatTypes
+        S = Tuple{Complex{FT}, FT}
+        array = rand(FT, 1, 3)
+
+        data = F{S}(array)
+        @test getfield(data, :array) == @view(array[:, :])
+
+        # test tuple assignment
+        data[1] = (Complex{FT}(-1.0, -2.0), FT(-3.0))
```
Comment (simonbyrne) [inline test/DataLayouts/data0d.jl:19]: it should act like a `Ref`:
```suggestion
        data[] = (Complex{FT}(-1.0, -2.0), FT(-3.0))
```

Comment (jb-mackay) [reply test/DataLayouts/data0d.jl:19]: Thank you! Was unaware of this. There are likely going to be more things like this that I miss as I build this PR.

File: src/DataLayouts/DataLayouts.jl (hunk 2 above)
Comment (jb-mackay) [reply src/DataLayouts/DataLayouts.jl]: Similarly for `Base.setindex!`?

Comment (simonbyrne) [reply src/DataLayouts/DataLayouts.jl]: Yes

File: src/DataLayouts/DataLayouts.jl [hunk 4]
```diff
@@ -411,6 +415,100 @@ end
...
+#     I::CartesianIndex{5},
+# )
+#     col[I[4]]
+# end
+@propagate_inbounds function Base.getindex(
+    col::Data0D,
+    I::CartesianIndex{5},
+)
+    col[I[4]]
+end
```
Comment (jb-mackay) [inline src/DataLayouts/DataLayouts.jl:488]: @simonbyrne can you tell me about these methods for get & set index? This is a place where I more or less have blindly copied.

File: src/DataLayouts/broadcast.jl [hunk 5]
```diff
@@ -3,6 +3,11 @@
 
 abstract type DataStyle <: Base.BroadcastStyle end
 
+abstract type Data0DStyle <: DataStyle end
+struct DataFStyle{A} <: Data0DStyle end
+DataStyle(::Type{DataF{S, A}}) where {S, A} = DataFStyle{parent_array_type(A)}()
+Data0DStyle(::Type{DataFStyle{A}}) where {A} = DataFStyle{A}
```
Comment (jb-mackay) [inline src/DataLayouts/broadcast.jl:9]: @simonbyrne when does this `Data0DStyle` need to be defined for other data styles? I'm looking at 
```
DataColumnStyle(::Type{VIFHStyle{Ni, A}}) where {Ni, A} = VFStyle{A}
DataSlab1DStyle(::Type{VIFHStyle{Ni, A}}) where {Ni, A} = IFStyle{Ni, A}
```
and wondering when the analogous methods are needed in this case.

**Pull request merged**